### PR TITLE
Fix Issue #708: Save files with string content

### PR DESCRIPTION
**Pull request by LincolnPuzey**
Wraps file objects open in string mode so they produce bytes and can be passed to `obj.upload_fileobj()`

### CODE REVIEW

**Comment by LincolnPuzey** [comment]
This fixes #708, #843.

@jschneier Can I get a review on this?

You can see my new tests that reproduce the issue, fail in the build on commit c3d00d9.

The latest build is just flake8 failing which is also happening on master.

File: storages/backends/s3boto3.py
Comment (LincolnPuzey) [inline storages/backends/s3boto3.py:445]: This is the core part of the fix. This approach has the benefits of not making any extra read() or seek() calls to the file object.

File: storages/utils.py
Comment (LincolnPuzey) [inline storages/utils.py:126]: Currently any string files will be encoded to utf-8 when saved.

I'm not sure about this. It could try to look up the `file.encoding` and use that.

Comment (LincolnPuzey) [reply storages/utils.py:126]: I think that's a good idea given how the `encoding` property is documented here https://docs.python.org/3/library/io.html#io.TextIOBase.encoding

**Comment by LincolnPuzey** [comment]
@jschneier Can I get this reviewed? Most of the diff is in tests. The new tests in `S3Boto3StorageTestsWithMoto` reproduce the original error and show its fixed.

**Comment by LincolnPuzey** [comment]
@jschneier Still waiting for a review

**Comment by LincolnPuzey** [comment]
Also fixes #961

**Comment by wiktr** [comment]
@LincolnPuzey I see there are conflicts, could you please resolve them?
@jschneier please take a look at your earliest convenience. I think it's worth resolving this and related tickets.

I hit the described issue using versions:

```
boto3==1.16.56
botocore==1.19.56
django-storages==1.11.1
```

Would be lovely if django storages worked with s3 and boto3 out of the box, without workarounds.

**Comment by LincolnPuzey** [comment]
I'm happy to resolve the conflicts if @jschneier or another maintainer can indicate they're willing to review and merge the PR.

**Comment by pauloxnet** [comment]
Thanks for the PR , it fix a problem we have in our projects.
I can image all maintainers are busy, but I hope one of them can review this PR and then merge.

**Review by pauloxnet — APPROVED** [review]

**Comment by dvf** [comment]
Any news here folks? The docs use an example that breaks too:

image

**Comment by mertcangokgoz** [comment]
Any update?

**Comment by hugo-reveni** [comment]
@LincolnPuzey here is another one interested in this topic, as we're having problems with text files uploaded from Admin using django-import-export v3.2.0

Any merge prevision?

**Comment by jschneier** [comment]
Thanks, `moto` is a big improvement. Seems boto has grown several more features in the last few years that necessitated a few more tweaks.

**Comment by hugo-reveni** [comment]
Thanks a lot @jschneier, we will wait for the next release with this improvement 👏🏽

**Comment by jschneier** [comment]
No need to wait; I released it already.

**Comment by hugo-reveni** [comment]
@jschneier sorry I didn't find new releases in Github tags section like past ones but now I see it was in [here](https://pypi.org/project/django-storages/). Thanks!

**Comment by jschneier** [comment]
Ah I forgot to apply the tag, fixed. Thanks.

**Comment by dvf** [comment]
It's people like @jschneier that make OSS awesome. Thanks Josh ❤️

**Pull request merged**
